pkg/build/prometheus/metrics.go

```diff
@@ -0,0 +1,109 @@
+package prometheus
+
+// NOTE, this module explicitly attempts to avoid references to the build api types/structs
+// in an attempt to start breaking up our compile dependencies
+
+import (
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// BuildSubsystem is how this build controller is represented in
+// prometheus metrics.
+const BuildSubsystem = "build_controller"
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	buildsCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_counters",
+			Help:      "Counts total teriminal builds by phase, and in case of failures, reason",
+		},
+		[]string{"phase", "reason"},
+	)
+	buildsGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_gauges",
+			Help:      "Show current counts of non-terminal builds by phase",
+		},
+		[]string{"phase"},
+	)
+	buildsOverallHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_timing_histograms",
+			Help:      "Histogram of durations of builds in seconds that have exited by phase",
+		},
+		[]string{"phase"},
+	)
+	buildStagesHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_stage_timing_histograms",
+			Help:      "Histogram of durations of stages in milliseconds that have exited by stage name",
+		},
+		[]string{"stage"},
+	)
+	buildStageStepsHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_stage_step_timing_histograms",
+			Help:      "Histogram of durations of steps in milliseconds that have exited by stage name and step name",
+		},
+		[]string{"stage", "step"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(buildsCounter)
+	prometheus.MustRegister(buildsGauge)
+	prometheus.MustRegister(buildsOverallHistogram)
+	prometheus.MustRegister(buildStagesHistogram)
+	prometheus.MustRegister(buildStageStepsHistogram)
+}
+
+func IntializeGauges() {
+	// with the gauges, if they were not explicitly initialized to 0 before actual use, they got unpredictable starting values
+	buildsGauge.WithLabelValues("new").Set(0)
+	buildsGauge.WithLabelValues("pending").Set(0)
+	buildsGauge.WithLabelValues("running").Set(0)
+}
+
+// UpdateTerminalBuildsCounter increments the prometheus counter for terminal builds based on phase, and in the case
+// of build failures, reason for the failure
+func UpdateTerminalBuildsCounter(phase, reason string) {
+	// reasons for the non-failed phases are benign / empty strings ...
+	// so we don't distinguish them from failed builds
```

Comment:
My comment here with "benign" was trying to convey it was *NOT* a problem if the reason was empty ... I'll reword and attempt to make that clearer